util/rate_limiter.cc

```diff
@@ -139,148 +139,105 @@ void GenericRateLimiter::Request(int64_t bytes, const Env::IOPriority pri,
   Req r(bytes, &request_mutex_);
   queue_[pri].push_back(&r);
 
+  // A thread representing a queued request coordinates with other such threads.
+  // There are two main duties.
+  //
+  // (1) Waiting for the next refill time.
+  // (2) Refilling the bytes and granting requests.
+  //
+  // To simplify the implementation, the duties can be performed by any thread.
+  // (1) could in theory be performed by all threads, whereas (2) can only be
+  // performed by one thread per refill interval while holding the mutex. (2)
+  // may even be performed by a thread not involved in (1).
+  //
+  // We restrict the flexibility a bit to reduce unnecessary wakeups:
+  //
+  // - `waiting_` flag ensures only one thread performs (1). This prevents the
+  //   thundering herd problem at the next refill time. The remaining threads
+  //   wait on their condition variable with an unbounded duration -- thus we
+  //   must remember to notify them to ensure forward progress.
+  // - (1) is typically done by a thread at the front of a queue. This is
+  //   trivial when the queues are initially empty as the first choice that
+  //   arrives must be the only entry in its queue. When queues are initially
+  //   non-empty, we achieve this by having (2) notify a thread at the front
+  //   of a queue (preferring higher priority) to perform the next duty.
+  // - We do not require any additional wakeup for (2). Typically it will just
+  //   be done by the thread that finished (1).
+  //
+  // Combined, the second and third bullet points above suggest the refill/
+  // granting will typically be done by a request at the front of its queue.
+  // This is important because one wakeup is saved when a granted request
+  // happens to be in an already running thread.
+  //
+  // Note this nice property is not guaranteed in a few cases, however.
```

Comment:
Question:
Is the "nice property" mentioned here referring to the refill/granting will typically (a)"be done by a request at the front of its queue" OR (b) "one wakeup is saved when a granted request happens to be in an already running thread"?

(a)(b) don't seem to be exactly the same to me since even in the existing impl where (a) is always the case, (b) is not always true for the same reason you mentioned in L175.

And it seems to me that L179-183 describes when (a) is not guaranteed while L175 describes when (b) is not guaranteed.

Did I miss anything here?

Reply:
Suggestion: 
I feel that L172-183 can be moved to PR summary as coding design details, instead of in the coding comment. The reason is:
(1) for new people who are not exposed to the existing impl and try to follow the code here, I doubt whether they will wonder too much about whether the heavy-lifting thread is front-queued and whether the performance gain from the "front-queued" property always exists. 
(2) for people who are exposed to the existing impl and wonder about the change on "front-queued" property, I anticipate that they would go after git blame and PR summary instead.

Reply:
> Question:
> Is the "nice property" mentioned here referring to the refill/granting will typically (a)"be done by a request at the front of its queue" OR (b) "one wakeup is saved when a granted request happens to be in an already running thread"?

Let's see, I meant to refer to (b) (woken up thread granting itself).

> (a)(b) don't seem to be exactly the same to me since even in the existing impl where (a) is always the case, (b) is not always true for the same reason you mentioned in L175.
> 
> And it seems to me that L179-183 describes when (a) is not guaranteed while L175 describes when (b) is not guaranteed.
>
> Did I miss anything here?

I am sure it is not written in the clearest way, but L176-178 (the point introducing the sub-points in L179-L183) intends to exclude cases where (a) is true but (b) isn't. In particular the condition "some requests in that same queue are granted" intends to exclude the L175 case.

Reply:
> Let's see, I meant to refer to (b) (woken up thread granting itself).
Oh - now I can see how the following paragraph describes cases (in order) where (b) is not guaranteed! I think the paragraph L174-183 is clear after (b) is made clear to be the "nice property" in L172. I was a bit confused only because (a) can also be seen a nice property and I ended up reading (b) as a consequence of the nice property (a).

Suggestion (with bias): 
Would you mind adding a few words like - "Note this nice property of woken up thread granting itself ... " just to specify a little bit more?